src/Client.js

```diff
@@ -11,23 +13,61 @@ function sanitizeAPIKey(key) {
   return key === '' ? null : key
 }
 
+function getTimeoutLength(timeout = 0, timeoutMaxAllowed) {
+  return (timeout - timeoutMaxAllowed) < 0 ? timeout : timeoutMaxAllowed
+}
+
 export default class Client {
   constructor(config) {
     this.config = Object.assign({}, defaults, config)
   }
 
   recognizeURL(url, options = {}) {
-    return this.connection(true, options)
-      .sendJson('/remote/recognitions', { url })
-      .then(this.handleResponse)
+    return this.recognizeRequest(
+      (options) => this.connection(true, options).sendJson('/remote/recognitions', { url }),
+      options
+    )
   }
 
   recognizeImage(data, options = {}) {
-    const params = {}
-    if (options.public) { params.public = true }
-    return this.connection(true, Object.assign({}, options, { params }))
-      .sendBinary('/recognitions', data)
-      .then(this.handleResponse)
+    const params = { public: options.public }
+    const fullOptions = Object.assign({}, options, { params })
+
+    return this.recognizeRequest(
+      (options) => this.connection(true, options).sendBinary('/recognitions', data),
+      fullOptions
+    )
+  }
+
+  // Takes a request and timeout and checks if the recognize request
+  // should start the polling process and calls poll if positive
+  recognizeRequest(requestFunc, options) {
+    if (options.timeout > 100) {
+      throw new ScnnrError('Timeout time greater than 100 not allowed')
+    }
+
+    const timeoutForFirstRequest = getTimeoutLength(options.timeout, 25)
+    const opt = Object.assign({}, options, { timeout: timeoutForFirstRequest })
+    const request = requestFunc(opt)
+
+    return new Promise((resolve, reject) => {
+      request
+        .then(this.handleResponse)
+        .then(recognition => {
+          if ((options.timeout || 0) > 0 && !recognition.isFinished()) {
+            return poll({
+              requestFunc: (options) => this.fetch(recognition.id, options),
+              conditionChecker: (recognition) => recognition.isFinished(),
```

Comment:
Recognitions can have `error` state as well. And I think the promise should be rejected if the API returns`error` state. What do you think?

Reply:
Completely agree. I'll implement it in the `handleResponse` function.

Reply:
In regards to this. In case there is an error, is the HTTP code 200? If its not the promise will be rejected at the `Connection` level.

Reply:
Yes, the API can return a result with `error` state even if the HTTP code is 200.
The reason is we have two types of synchronous errors and asynchronous errors.

For synchronous errors, we can return 4xx response.
But for asynchronous errors, we return 202 response with the `queued` state at first.
And then we return 200 response with the `error` state for a GET request, because the GET request isn't wrong and we cannot use the HTTP code 4xx.

This is the current specification of the API. Am I making sense?